apps/tup-ui/src/App.tsx

```diff
@@ -1,8 +1,65 @@
 import { useState } from 'react';
+import { Outlet, Routes, Route, Navigate, useLocation } from 'react-router-dom';
 import { Message } from '@tacc/core-components';
+import { useAuth, useJwt } from './hooks';
 import LoginComponent from './auth/LoginComponent';
+import ProfileComponent from './auth/ProfileComponent';
+
+const AppLayout = () => {
+  const [count, setCount] = useState(0);
+  return (
+    <div className="App">
+      <header className="App-header">
+        <p>Hello Vite + React!</p>
+        <Message />
+        <p>
+          <button type="button" onClick={() => setCount((count) => count + 1)}>
+            count is: {count}
+          </button>
+        </p>
+        <Outlet />
+      </header>
+    </div>
+  );
+};
+
+function RequireAuth({ children }: { children: JSX.Element }) {
+  const { loggedIn } = useAuth();
+  const { isLoading } = useJwt();
+  const location = useLocation();
+
+  // React-query fetches the JWT asynchronously so we need to wait for it
+  // (this should be instantaneous).
+  if (isLoading) return null;
+
+  if (!loggedIn) {
+    // Redirect them to the /login page, but save the current location they were
+    // trying to go to when they were redirected. This allows us to send them
+    // along to that page after they login, which is a nicer user experience
+    // than dropping them off on the home page.
+    return <Navigate to="/login" state={{ from: location }} replace />;
+  }
+
+  return children;
+}
 
 function App() {
+  return (
+    <Routes>
+      <Route path="/" element={<AppLayout />}>
+        <Route
```

Comment:
I think in tapis-ui there was a `<ProtectedRoute>` element (https://github.com/tapis-project/tapis-ui/blob/main/src/tapis-ui/_common/ProtectedRoute/ProtectedRoute.tsx), rather than using Route and then wrapping the element in the Require. Would that allow us to avoid wrapping every route's elements in `<RequireAuth>`?

Reply:
I thought about this. The tradeoff is that the `ProtectedRoute` component needs to inherit `Route`'s interface and I was afraid that might get hairy. I'll see if this actually simplifies things. FWIW the route guard code comes straight from the [docs](https://reactrouter.com/docs/en/v6/examples/auth) which suggest a `<ProtectedRoute>` in v5 and this approach in v6.

Reply:
OK, this is actually pretty hairy. Trying to wrap a `<Route/>` as `<ProtectedRoute/>` gives this error: 

`[ProtectedRoute] is not a <Route> component. All component children of <Routes> must be a <Route> or <React.Fragment>`

Seems like React-Router v6 very specifically disallows this pattern.